File: cpp/TextNormalizer.cpp
#include "TextNormalizer.h"
#include <regex>
#include <algorithm>

TextNormalizer::TextNormalizer() {
    // Constructor
}

TextNormalizer::~TextNormalizer() {
    // Destructor
}

std::string TextNormalizer::normalizeText(const std::string& input) {
    if (input.empty()) {
        return input;
    }

    std::string result = input;
    
    // Normalize stretched words (e.g., "hellooo" -> "hello")
    result = normalizeRepeatedCharacters(result);
    
    // Remove text in asterisks (e.g., "*cough*")
    result = removeAsterisks(result);
    
    // Normalize capitalization (e.g., "HeLLo" -> "Hello")
    result = normalizeCasing(result);
    
    return result;
}

std::string TextNormalizer::normalizeRepeatedCharacters(const std::string& input) {
    if (input.empty()) {
        return input;
    }

    // Define word boundaries to identify words
    std::regex wordPattern("\\b\\w+\\b");
    
    std::string result;
    std::sregex_iterator wordIterBegin(input.begin(), input.end(), wordPattern);
    std::sregex_iterator wordIterEnd;
    
    size_t lastPos = 0;
    for (std::sregex_iterator iter = wordIterBegin; iter != wordIterEnd; ++iter) {
        std::smatch match = *iter;
        std::string word = match.str();
        
        // Add text before the current word
        result += input.substr(lastPos, match.position() - lastPos);
        
        // Normalize the word
        std::string normalized;
        for (size_t i = 0; i < word.length(); i++) {
            char current = word[i];
            normalized += current;
            
            // Skip all subsequent identical characters
            while (i + 1 < word.length() && word[i + 1] == current) {
                i++;
            }
        }
        
        result += normalized;
        lastPos = match.position() + match.length();
    }
    
    // Add remaining text
    result += input.substr(lastPos);
    
    return result;
}

std::string TextNormalizer::removeAsterisks(const std::string& input) {
    if (input.empty()) {
        return input;
    }
    
    // Remove text between asterisks completely
    std::regex asteriskPattern("\\*[^*]*\\*");
    std::string result = std::regex_replace(input, asteriskPattern, "");
    
    // Fix any resulting double spaces or more
    std::regex multipleSpacesPattern("\\s{2,}");
    result = std::regex_replace(result, multipleSpacesPattern, " ");
    
    return result;
}

std::string TextNormalizer::normalizeCasing(const std::string& input) {
    if (input.empty()) {
        return input;
    }

    // Define word boundaries to identify words
    std::regex wordPattern("\\b\\w+\\b");
    
    std::string result;
    std::sregex_iterator wordIterBegin(input.begin(), input.end(), wordPattern);
    std::sregex_iterator wordIterEnd;
    
    size_t lastPos = 0;
    for (std::sregex_iterator iter = wordIterBegin; iter != wordIterEnd; ++iter) {
        std::smatch match = *iter;
        std::string word = match.str();
        
        // Add text before the current word
        result += input.substr(lastPos, match.position() - lastPos);
        
        // If the word has mixed capitalization
        if (hasMixedCasing(word)) {
            // Convert everything to lowercase
            std::string normalized = word;
            std::transform(normalized.begin(), normalized.end(), normalized.begin(), ::tolower);
            
            // First letter uppercase if the original was uppercase
            if (std::isupper(word[0])) {
                normalized[0] = std::toupper(normalized[0]);
            }
            
            result += normalized;
        } else {
            result += word;
        }
        
        lastPos = match.position() + match.length();
    }
    
    // Add remaining text
    result += input.substr(lastPos);
    
    return result;
}

bool TextNormalizer::hasMixedCasing(const std::string& word) {
    if (word.length() <= 1) {
        return false;
    }
    
    bool hasUpper = false;
    bool hasLower = false;
    
    for (char c : word) {
        if (std::isupper(c)) {
            hasUpper = true;
        } else if (std::islower(c)) {
            hasLower = true;
        }
        
        if (hasUpper && hasLower) {
            return true;
        }
    }
    
    return false;
}

File: r6/text_normalizer.R
#' TextNormalizer Class
#' 
#' A class for normalizing text containing unusual writing styles,
#' such as repeated characters, text in asterisks, or
#' unusual capitalization.
#' 
#' @importFrom stringr str_replace_all str_detect
#' @examples
#' normalizer <- TextNormalizer$new()
#' normalizer$normalize_text("Hellooo how areee you *cough* today?")
TextNormalizer <- R6::R6Class(
  "TextNormalizer",
  public = list(
    #' @description
    #' Normalize text by applying multiple normalization rules.
    #' @param input The text to normalize.
    #' @return The normalized text.
    normalize_text = function(input) {
      if (is.null(input) || input == "") {
        return(input)
      }
      
      result <- input
      
      # Normalize stretched words (e.g., "hellooo" -> "hello")
      result <- self$normalize_repeated_characters(result)
      
      # Remove text in asterisks (e.g., "*cough*")
      result <- self$remove_asterisks(result)
      
      # Normalize capitalization (e.g., "HeLLo" -> "Hello")
      result <- self$normalize_casing(result)
      
      return(result)
    },
    
    #' @description
    #' Normalize repeated characters in text.
    #' @param input The text to normalize.
    #' @return The text with normalized repeated characters.
    normalize_repeated_characters = function(input) {
      if (is.null(input) || input == "") {
        return(input)
      }
      
      # Split by words and process each word
      words <- unlist(stringr::str_extract_all(input, "\\b\\w+\\b"))
      
      for (word in words) {
        normalized_word <- ""
        i <- 1
        
        while (i <= nchar(word)) {
          current <- substr(word, i, i)
          normalized_word <- paste0(normalized_word, current)
          
          # Skip all subsequent identical characters
          while (i + 1 <= nchar(word) && substr(word, i + 1, i + 1) == current) {
            i <- i + 1
          }
          
          i <- i + 1
        }
        
        # Replace the original word with the normalized one
        input <- stringr::str_replace_all(input, fixed(word), normalized_word)
      }
      
      return(input)
    },
    
    #' @description
    #' Remove text surrounded by asterisks and fixes any resulting extra spaces.
    #' @param input The text to normalize.
    #' @return The text without asterisk content and no extra spaces.
    remove_asterisks = function(input) {
      if (is.null(input) || input == "") {
        return(input)
      }
      
      # Remove text between asterisks completely
      result <- stringr::str_replace_all(input, "\\*[^*]*\\*", "")
      
      # Fix any resulting double spaces or more
      result <- stringr::str_replace_all(result, "\\s{2,}", " ")
      
      return(result)
    },
    
    #' @description
    #' Normalize capitalization of words.
    #' @param input The text to normalize.
    #' @return The text with normalized capitalization.
    normalize_casing = function(input) {
      if (is.null(input) || input == "") {
        return(input)
      }
      
      # Split by words and process each word
      words <- unlist(stringr::str_extract_all(input, "\\b\\w+\\b"))
      
      for (word in words) {
        if (self$has_mixed_casing(word)) {
          # Convert everything to lowercase
          normalized <- tolower(word)
          
          # First letter uppercase if the original was uppercase
          if (stringr::str_detect(substr(word, 1, 1), "[A-Z]")) {
            normalized <- paste0(toupper(substr(normalized, 1, 1)), 
                                substr(normalized, 2, nchar(normalized)))
          }
          
          # Replace the original word with the normalized one
          input <- stringr::str_replace_all(input, fixed(word), normalized)
        }
      }
      
      return(input)
    },
    
    #' @description
    #' Check if a word contains mixed capitalization.
    #' @param word The word to check.
    #' @return TRUE if the word has mixed capitalization, otherwise FALSE.
    has_mixed_casing = function(word) {
      if (is.null(word) || nchar(word) <= 1) {
        return(FALSE)
      }
      
      has_upper <- FALSE
      has_lower <- FALSE
      
      for (i in 1:nchar(word)) {
        char <- substr(word, i, i)
        
        if (stringr::str_detect(char, "[A-Z]")) {
          has_upper <- TRUE
        } else if (stringr::str_detect(char, "[a-z]")) {
          has_lower <- TRUE
        }
        
        if (has_upper && has_lower) {
          return(TRUE)
        }
      }
      
      return(FALSE)
    }
  )
)

# Example usage
# library(R6)
# 
# normalizer <- TextNormalizer$new()
# 
# # Example for text preprocessing
# text_input <- "Hellooo how areee you *cough* today? *laughs* It's a niiice day!"
# normalized_text <- normalizer$normalize_text(text_input)
# cat("Original input:", text_input, "\n")
# cat("Normalized text:", normalized_text, "\n")
# 
# # Example with mixed capitalization
# mixed_case_input <- "HeLLo ThERe!!! Hooooow are you??"
# normalized_case <- normalizer$normalize_text(mixed_case_input)
# cat("Original input:", mixed_case_input, "\n")
# cat("Normalized text:", normalized_case, "\n")